README.md
# web_mercator

Dart implementation of the uber [@math.gl/web-mercator](https://github.com/uber-web/math.gl/tree/master/modules/web-mercator) javascript module.

## Getting Started

See the `tests` for in-depth implementation methods.
Here is a basic example using the `fitBounds` factory with end user device informations:
```dart
import 'package:web_mercator/web_mercator.dart' show MercatorViewport

final viewport = MercatorViewport.fitBounds(width: deviceWidth, height: deviceHeight, bounds: northEastSoutWest);
```

With that viewport, you can now project coordinates passing a `Vector` to the `project` method. Using a `Vector2` will result in a flat projection (onto a sphere) while a `Vector3` will consider the z component as an altidude expressed in meters.

The `example/` folder contains a sample app displaying the boudary of Martinique with interactive pitch, bearing & zoom touch interactions.

CHANGELOG.md
## [0.9.0] - 2020/01/26

* Initial release covering most of the relevant bits.

## [0.9.1-dev.1] - 2020/01/26

* Improving publishing score.

## [0.9.1] - 2020/01/26

* Passing relevant attributes of package scoring.

## [0.9.2] - 2020/01/30

* Add a `bbox` function to facilitate projection from a list of coordinates.

## [0.9.3] - 2020/02/11

* Change the name of `Viewport` to `MercatorViewport` so it doesn't clash with the flutter's [Viewport](https://api.flutter.dev/flutter/widgets/Viewport-class.html) class.
* Strongly typing some variables as suggested by dart linter.
* Add an example.

## [0.9.4] - 2020/03/20

* Add a convenient `copyWith` factory to clone a viewport with parameters typically reflecting user interactions.
* Update the example accordingly…

## [0.9.4+1] - 2020/06/02

* Cosmetic changes tackling analysis updates.

test/samples.dart
class ViewportData {
  final String name;
  final double width, height;
  final double lng, lat;
  final double zoom, bearing;
  final double pitch, altitude;

  const ViewportData._(this.name, this.width, this.height, this.lng, this.lat, this.zoom, this.bearing, {this.pitch, this.altitude});

  factory ViewportData.flat() => ViewportData._('Flat', 800, 600, -122.43, 37.75, 11.5, 0);
  factory ViewportData.pitched() => ViewportData._('Pitched', 800, 600, -122.43, 37.75, 11.5, 0, pitch: 30);
  factory ViewportData.rotated() => ViewportData._('Rotated', 1267, 400, -122.4194, 37.7749, 11, 180, pitch: 60, altitude: 1.5);
  factory ViewportData.highLatitude() => ViewportData._('HighLatitude', 500, 500, 42.42694, 75.751537, 15.5, -40, pitch: 20, altitude: 1.5);

  factory ViewportData.copyWith(ViewportData original, {double lng, double lat}) => ViewportData._(
        original.name,
        original.width,
        original.height,
        lng ?? original.lng,
        lat ?? original.lat,
        original.zoom,
        original.bearing,
        pitch: original.pitch,
        altitude: original.altitude,
      );

  @override
  String toString() => name;
  // - width: $width, height: $height
  // - lng: $lng, lat: $lat
  // - zoom: $zoom, bearing: $bearing
  // - pitch: ${pitch ?? 'nan'}, altitude: ${altitude ?? 'nan'}''';
}

List<ViewportData> get viewports => [ViewportData.flat(), ViewportData.pitched(), ViewportData.rotated(), ViewportData.highLatitude()];

dynamic get bounds => [
      [
        {
          'width': 100,
          'height': 100,
          // southwest bound first
          'bounds': [
            [-73.9876, 40.7661],
            [-72.9876, 41.7661]
          ]
        },
        {'lng': -73.48759999999997, 'lat': 41.26801443944763, 'zoom': 5.723804361273887}
      ],
      [
        {
          'width': 100,
          'height': 100,
          // northeast bound first
          'bounds': [
            [-72.9876, 41.7661],
            [-73.9876, 40.7661]
          ]
        },
        {'lng': -73.48759999999997, 'lat': 41.26801443944763, 'zoom': 5.723804361273887}
      ],
      [
        {
          'width': 100,
          'height': 100,
          'bounds': [
            [-73.0, 10.0],
            [-73.0, 10.0]
          ],
          'maxZoom': 22.0
        },
        {'lng': -73, 'lat': 10, 'zoom': 22}
      ],
      [
        {
          'width': 100,
          'height': 100,
          'bounds': [
            [-73.0, 10.0],
            [-73.0, 10.0]
          ],
          'minExtent': 0.01
        },
        {'lng': -73, 'lat': 10, 'zoom': 13.28771238}
      ],
      [
        {
          'width': 600,
          'height': 400,
          'bounds': [
            [-23.407, 64.863],
            [-23.406, 64.874]
          ],
          'padding': 20,
          'offset': [0, -40]
        },
        {'lng': -23.406499999999973, 'lat': 64.86850056273362, 'zoom': 12.89199533073045}
      ],
      [
        {
          'width': 600,
          'height': 400,
          'bounds': [
            [-23.407, 64.863],
            [-23.406, 64.874]
          ],
          'padding': {'top': 100, 'bottom': 10, 'left': 30, 'right': 30},
          'offset': [0, -40]
        },
        {'lng': -23.406499999999973, 'lat': 64.870857602, 'zoom': 12.476957831}
      ]
    ];

dynamic get projections => [
      {
        'title': 'project (center)',
        'func': 'project',
        'input': [-122.43, 37.75],
        'expected': [400, 300]
      },
      {
        'title': 'unproject (center)',
        'func': 'unproject',
        'input': [400.0, 300.0],
        'expected': [-122.43, 37.75]
      },
      {
        'title': 'project (corner)',
        'func': 'project',
        'input': [-122.55, 37.83],
        'expected': [-1.329741801625046, 6.796120915775314]
      },
      {
        'title': 'unproject (corner)',
        'func': 'unproject',
        'input': [.0, .0],
        'expected': [-122.55024809579456, 37.832294933238586]
      }
    ];


test/viewport_samples.dart
class ViewportData {
  final int width, height;
  final double lng, lat, zoom, bearing;
  final double pitch, altitude;

  ViewportData(this.width, this.height, this.lng, this.lat, this.zoom, this.bearing, {this.pitch, this.altitude});

  @override
  String toString() => '''
  - width: $width, height: $height
  - lng: $lng, lat: $lat
  - zoom: $zoom, bearing: $bearing
  - pitch: ${pitch ?? 'nan'}, altitude: ${altitude ?? 'nan'}''';
}

class FlatViewport extends ViewportData {
  FlatViewport() : super(800, 600, -122.43, 37.75, 11.5, 0);

  @override
  String toString() => 'FlatViewport';
}

class PitchedViewport extends ViewportData {
  PitchedViewport() : super(800, 600, -122.43, 37.75, 11.5, 0, pitch: 30);

  @override
  String toString() => 'PitchedViewport';
}

class RotatedViewport extends ViewportData {
  RotatedViewport() : super(1267, 400, -122.4194, 37.7749, 11, 180, pitch: 60, altitude: 1.5);

  @override
  String toString() => 'RotatedViewport';
}

class HighLatitudeViewport extends ViewportData {
  HighLatitudeViewport() : super(500, 500, 42.42694, 75.751537, 15.5, -40, pitch: 20, altitude: 1.5);

  @override
  String toString() => 'HighLatitudeViewport';
}

List<ViewportData> get data => [FlatViewport(), PitchedViewport(), RotatedViewport(), HighLatitudeViewport()];


lib/utils.dart
part of web_mercator;

// CONSTANTS
const PI_2 = pi * .5;
const PI_4 = pi * .25;
const TILE_SIZE = 512;

/// Average circumference (40075 km equatorial, 40007 km meridional)
const EARTH_CIRCUMFERENCE = 40.03e6;

/// Mapbox default altitude
const DEFAULT_ALTITUDE = 1.5;

// Util functions
num log2(num x) => log(x) / log(2);
num zoomToScale(double zoom) => pow(2, zoom);
num scaleToZoom(double scale) => log2(scale);

Matrix4 perspective({
  double fovy,
  double aspect,
  double near,
  double far,
}) {
  final out = Matrix4.zero();
  final f = 1.0 / tan(fovy / 2);

  out[0] = f / aspect;
  out[5] = f;
  out[11] = -1;

  if (far != null && !far.isInfinite) {
    final nf = 1 / (near - far);
    out[10] = (far + near) * nf;
    out[14] = (2 * far * near) * nf;
  } else {
    out[10] = -1;
    out[14] = -2 * near;
  }

  return out;
}

/// Transforms a [vector] with a projection [matrix].
Vector4 transformVector(Vector4 vector, Matrix4 matrix) {
  final res = vector.clone();
  res.applyMatrix4(matrix);
  return res.scaled(1 / res[3]);
}

/// Project [lng, lat] on sphere onto [x, y] on 512*512 Mercator Zoom 0 tile.
/// Performs the nonlinear part of the web mercator projection.
/// Remaining projection is done with 4x4 matrices which also handles perspective.
Vector2 lngLatToWorld(double lng, double lat) {
  assert(lng.isFinite);
  assert(lat.isFinite && lat >= -90 && lat <= 90, 'invalid lat');

  final lambda2 = lng * degrees2Radians;
  final phi2 = lat * degrees2Radians;
  final x = (TILE_SIZE * (lambda2 + pi)) / (2 * pi);
  final y = (TILE_SIZE * (pi + log(tan(PI_4 + phi2 * 0.5)))) / (2 * pi);

  return Vector2(x, y);
}

/// Unproject world point [x, y] on map onto [lng, lat] on sphere.
Vector2 worldToLngLat(double x, double y) {
  final lambda2 = (x / TILE_SIZE) * (2 * pi) - pi;
  final phi2 = 2 * (atan(exp((y / TILE_SIZE) * (2 * pi) - pi)) - PI_4);
  return Vector2(lambda2 * radians2Degrees, phi2 * radians2Degrees);
}

/// Returns the zoom level that gives a 1 meter pixel at a certain [lat].
double getMeterZoom(double lat) {
  assert(lat.isFinite);
  final latCosine = cos(lat * degrees2Radians);
  return scaleToZoom(EARTH_CIRCUMFERENCE * latCosine) - 9;
}

/// Calculate distance scales in meters around current [lat, lng], both for degrees and pixels.
Map<String, List<double>> getDistanceScales(lng, lat, {highPrecision = false}) {
  assert(lng.isFinite && lat.isFinite);

  final latCosine = cos(lat * degrees2Radians);

  /// Number of pixels occupied by one degree lng around current lat/lon:
  const unitsPerDegreeX = TILE_SIZE / 360;
  final unitsPerDegreeY = unitsPerDegreeX / latCosine;

  /// Number of pixels occupied by one meter around current lat/lon:
  final altUnitsPerMeter = TILE_SIZE / EARTH_CIRCUMFERENCE / latCosine;

  final result = Map();
  result['unitsPerMeter'] = [altUnitsPerMeter, altUnitsPerMeter, altUnitsPerMeter];
  result['metersPerUnit'] = [1 / altUnitsPerMeter, 1 / altUnitsPerMeter, 1 / altUnitsPerMeter];
  result['unitsPerDegree'] = [unitsPerDegreeX, unitsPerDegreeY, altUnitsPerMeter];
  result['degreesPerUnit'] = [1 / unitsPerDegreeX, 1 / unitsPerDegreeY, 1 / altUnitsPerMeter];

  if (highPrecision) {
    final latCosine2 = (degrees2Radians * tan(lat * degrees2Radians)) / latCosine;
    final unitsPerDegreeY2 = (unitsPerDegreeX * latCosine2) * .5;
    final altUnitsPerDegree2 = (TILE_SIZE / EARTH_CIRCUMFERENCE) * latCosine2;
    final altUnitsPerMeter2 = (altUnitsPerDegree2 / unitsPerDegreeY) * altUnitsPerMeter;

    result['unitsPerDegree2'] = [0, unitsPerDegreeY2, altUnitsPerDegree2];
    result['unitsPerMeter2'] = [altUnitsPerMeter2, 0, altUnitsPerMeter2];
  }

  return result;
}

Map<String, double> getProjParameters({
  int width,
  int height,
  double altitude = DEFAULT_ALTITUDE,
  double pitch = 0,
  double nearZMultiplier = 1,
  double farZMultiplier = 1,
}) {
  /// Find the distance from the center point to the center top in altitude units using law of sines.
  final pitchRadians = pitch * degrees2Radians;
  final halfFov = atan(.5 / altitude);
  final topHalfSurfaceDistance = (sin(halfFov) * altitude) / sin(PI_2 - pitchRadians - halfFov);

  /// Calculate z value of the farthest fragment that should be rendered.
  final farZ = cos(PI_2 - pitchRadians) * topHalfSurfaceDistance + altitude;

  return {
    'fov': 2 * halfFov,
    'aspect': width / height,
    'focalDistance': altitude,
    'near': nearZMultiplier,
    'far': farZ * farZMultiplier,
  };
}

/// Offset a [lngLatZ] position by meterOffset (northing, easting) [xyz].
Vector3 addMetersToLngLat(Vector3 lngLatZ, Vector3 xyz) {
  final lng = lngLatZ[0], lat = lngLatZ[1], z0 = lngLatZ[3];
  final x = xyz[0], y = xyz[1], z = xyz[2];

  final distanceScales = getDistanceScales(lat, lng, highPrecision: true);
  final unitsPerMeter = distanceScales['unitsPerMeter'];
  final unitsPerMeter2 = distanceScales['unitsPerMeter2'];

  final worldspace = lngLatToWorld(lng, lat);
  worldspace[0] += x * (unitsPerMeter[0] + unitsPerMeter2[0] * y);
  worldspace[1] += y * (unitsPerMeter[1] + unitsPerMeter2[1] * y);

  final newLngLat = worldToLngLat(worldspace[0], worldspace[1]);
  final newZ = (z0 ?? 0) + (z ?? 0);

  return z.isFinite || z0.isFinite ? Vector3(newLngLat[0], newLngLat[1], newZ) : newLngLat;
}

Matrix4 getViewMatrix({
  int height,
  double pitch,
  double bearing,
  double altitude,
  double scale,
  Vector2 center,
}) {
  scale /= height;
  final vm = Matrix4.identity()
    ..translate(0, 0, -altitude)
    ..rotateX(-pitch * -pitch * degrees2Radians)
    ..rotateZ(bearing * degrees2Radians)
    ..scale(scale, scale, scale);

  if (center != null) {
    center.negate();
    vm.translate(center);
  }

  return vm;
}

Matrix4 getProjMatrix({
  int width,
  int height,
  double pitch,
  double altitude,
  double nearZMultiplier,
  double farZMultiplier,
}) {
  final projParams = getProjParameters(
    width: width,
    height: height,
    altitude: altitude,
    pitch: pitch,
    nearZMultiplier: nearZMultiplier,
    farZMultiplier: farZMultiplier,
  );

  return perspective(
    fovy: projParams['fov'],
    aspect: projParams['aspect'],
    near: projParams['near'],
    far: projParams['far'],
  );
}

/// Project flat coordinates [xyz] to pixels on screen given the [pixelProjMatrix].
Vector4 worldToPixels(Vector3 xyz, Matrix4 pixelProjMatrix) {
  final x = xyz[0], y = xyz[1], z = xyz[2];
  assert(x.isFinite && y.isFinite && z.isFinite);

  return transformVector(Vector4(x, y, z, 1), pixelProjMatrix);
}

/// Unproject [xyz] pixels on screen to flat coordinates given the [pixelUnprojMatrix].
Vector pixelsToWorld(Vector3 xyz, Matrix4 pixelUnprojMatrix, {double targetZ}) {
  final x = xyz[0], y = xyz[1], z = xyz[2];
  assert(x.isFinite && y.isFinite, 'invalid pixel coordinate');

  if (z.isFinite) {
    final coord = transformVector(Vector4(x, y, z, 1), pixelUnprojMatrix);
    return coord;
  }

  /// since we don't know the correct projected z value for the point, unproject two points to get a line and then find the point on that line with z=0
  final coord0 = transformVector(Vector4(x, y, 0, 1), pixelUnprojMatrix);
  final coord1 = transformVector(Vector4(x, y, 1, 1), pixelUnprojMatrix);

  final z0 = coord0[2], z1 = coord1[2];
  final t = z0 == z1 ? 0 : ((targetZ ?? 0) - z0) / (z1 - z0);

  /// lerp
  final ax = coord0[0], ay = coord0[1];
  return Vector2(ax + t * (coord1[0] - ax), ay + t * (coord1[1] - ay));
}


lib/viewport.dart
part of web_mercator;

class Viewport {
  final num width, height;
  final double lat, lng, zoom, pitch, bearing, altitude, unitsPerMeter;
  final Vector2 center;

  Matrix4 viewMatrix, projMatrix;
  Matrix4 _viewProjMatrix, _pixelProjMatrix, _pixelUnprojMatrix;

  Viewport({
    @required width,
    @required height,
    this.lng = .0,
    this.lat = .0,
    this.zoom = .0,
    this.pitch = .0,
    this.bearing = .0,
    this.altitude = 1.5,
    double nearZMultiplier = .02,
    double farZMultiplier = 1.01,
  })  : assert(altitude >= .75, 'invalid altitude'),
        this.width = max(1, width),
        this.height = max(1, height),
        unitsPerMeter = getDistanceScales(lng, lat)['unitsPerMeter'][2],
        center = lngLatToWorld(lng, lat) {
    viewMatrix = getViewMatrix(
      height: this.height,
      pitch: pitch,
      bearing: bearing,
      altitude: max(.75, altitude),
      scale: zoomToScale(zoom),
      center: center,
    );

    projMatrix = getProjMatrix(
      width: this.width,
      height: this.height,
      pitch: pitch,
      altitude: altitude,
      nearZMultiplier: nearZMultiplier,
      farZMultiplier: farZMultiplier,
    );

    _viewProjMatrix = Matrix4.identity()..multiply(projMatrix)..multiply(viewMatrix);

    _pixelProjMatrix = Matrix4.identity()
      ..scale(this.width * .5, -this.height * .5, 1)
      ..translate(1.0, -1.0, .0)
      ..multiply(_viewProjMatrix);

    _pixelUnprojMatrix = Matrix4.inverted(_pixelProjMatrix);
  }

  /// Returns a new viewport that fit around the given rectangle.
  factory Viewport.fitBounds({
    @required int width,
    @required int height,
    @required List<num> bounds,
    double minExtent = 0,
    double maxZoom = 24,
    dynamic padding = 0,
    List<num> offset = const [0, 0],
  }) {
    final lngLatZoom = fitBounds(
      width: width,
      height: height,
      bounds: bounds,
      minExtent: minExtent,
      maxZoom: maxZoom,
      padding: padding,
      offset: offset,
    );

    return Viewport(width: width, height: height, lng: lngLatZoom['lng'], lat: lngLatZoom['lat'], zoom: lngLatZoom['zoom']);
  }

  /// Project [vector] to pixel coordinates.
  Vector project(Vector vector, {bool topLeft = true}) {
    assert(vector is Vector2 || vector is Vector3);

    final worldPosition = projectPosition(vector);
    final coord = worldToPixels(worldPosition, _pixelProjMatrix);
    final y = topLeft ? coord[1] : height - coord[1];

    return vector is Vector2 ? Vector2(coord[0], y) : Vector3(coord[0], y, coord[2]);
  }

  ///  Unproject [xyz] coordinates onto world coordinates.
  Vector unproject(Vector xyz, {bool topLeft = true, double targetZ}) {
    assert(xyz is Vector2 || xyz is Vector3);

    var vec, z;
    try {
      vec = xyz as Vector2;
      z = double.nan;
    } on CastError {
      vec = xyz as Vector3;
      z = vec[2];
    }

    final coord = pixelsToWorld(
      Vector3(vec[0], topLeft ? vec[1] : height - vec[1], z),
      _pixelUnprojMatrix,
      targetZ: targetZ != null ? targetZ * unitsPerMeter : null,
    );

    final unprojPosition = unprojectPosition(coord);
    if (vec is Vector3) {
      return unprojPosition;
    } else if (targetZ != null) {
      return Vector3(unprojPosition[0], unprojPosition[1], targetZ);
    } else {
      return Vector2(unprojPosition[0], unprojPosition[1]);
    }
  }

  Vector3 projectPosition(Vector vector) {
    assert(vector is Vector2 || vector is Vector3 || vector is Vector4);

    var vec;
    try {
      vec = vector as Vector2;
    } on CastError {
      try {
        vec = vector as Vector3;
      } on CastError {
        vec = vector as Vector4;
      }
    }

    final flatProjection = projectFlat(vec[0], vec[1]);
    final z = (vector is Vector3 ? vector[2] : 0) * unitsPerMeter;

    return Vector3(flatProjection[0], flatProjection[1], z);
  }

  Vector3 unprojectPosition(Vector vector) {
    assert(vector is Vector2 || vector is Vector3 || vector is Vector4);

    var vec;
    try {
      vec = vector as Vector2;
    } on CastError {
      try {
        vec = vector as Vector3;
      } on CastError {
        vec = vector as Vector4;
      }
    }

    final unprojection = unprojectFlat(vec[0], vec[1]);
    final z = (vec is Vector3 || vec is Vector4 ? vec[2] : 0) / unitsPerMeter;

    return Vector3(unprojection[0], unprojection[1], z);
  }

  Vector2 projectFlat(double x, double y) => lngLatToWorld(x, y);

  Vector2 unprojectFlat(double x, double y) => worldToLngLat(x, y);

  /// Get the map center that places a given [lngLat] coordinate at screen point [pos].
  Vector2 getLocationAtPoint({Vector2 lngLat, Vector2 pos}) {
    final fromLocation = pixelsToWorld(Vector3(pos[0], pos[1], double.nan), _pixelUnprojMatrix) as Vector2;
    final toLocation = lngLatToWorld(lngLat[0], lngLat[1]);

    print('fromLocation: $fromLocation, toLocation: $toLocation');

    final translate = toLocation.clone();
    fromLocation.negate();
    translate.add(fromLocation);

    final newCenter = center.clone();
    newCenter.add(translate);

    print('translate: $translate, newCenter: $newCenter');
    return worldToLngLat(newCenter[0], newCenter[1]);
  }

  @override
  int get hashCode => quiver.hashObjects([width, height, viewMatrix, projMatrix]);

  @override
  bool operator ==(Object other) => (other is Viewport) && (other.width == this.width && other.height == this.height && other.viewMatrix == this.viewMatrix);
}


lib/math_utils.dart
part of web_mercator;

const PI_2 = pi * .5;
const PI_4 = pi * .25;

num log2(num x) => log(x) / log(2);
num zoomToScale(double zoom) => pow(2, zoom);
num scaleToZoom(double scale) => log2(scale);

/// Generates a perspective projection matrix with the given bounds.
/// see: [gl-matrix/mat4.js](https://github.com/toji/gl-matrix/blob/master/src/mat4.js)
Matrix4 perspective({double fovy, double aspect, double near, double far}) {
  final out = Matrix4.zero();
  final f = 1.0 / tan(fovy / 2);

  out[0] = f / aspect;
  out[5] = f;
  out[11] = -1;

  if (far != null && !far.isInfinite) {
    final nf = 1 / (near - far);
    out[10] = (far + near) * nf;
    out[14] = (2 * far * near) * nf;
  } else {
    out[10] = -1;
    out[14] = -2 * near;
  }

  return out;
}

/// Transforms a [vector] with a projection [matrix].
Vector4 transformVector(Vector4 vector, Matrix4 matrix) {
  final res = vector.clone();
  res.applyMatrix4(matrix);
  return res.scaled(1 / res[3]);
}

Matrix4 getViewMatrix({
  num height,
  double pitch,
  double bearing,
  double altitude,
  double scale,
  Vector2 center,
}) {
  scale /= height;
  final vm = Matrix4.identity()
    ..translate(.0, .0, -altitude)
    ..rotateX(-pitch * degrees2Radians)
    ..rotateZ(bearing * degrees2Radians)
    ..scale(scale, scale, scale);

  if (center != null) {
    final centerClone = center.clone();
    centerClone.negate();
    vm.translate(centerClone[0], centerClone[1], .0);
  }

  return vm;
}

Matrix4 getProjMatrix({
  num width,
  num height,
  double pitch,
  double altitude,
  double nearZMultiplier,
  double farZMultiplier,
}) {
  final projParams = getProjParameters(
    width,
    height,
    altitude: altitude,
    pitch: pitch,
    nearZMultiplier: nearZMultiplier,
    farZMultiplier: farZMultiplier,
  );

  return perspective(
    fovy: projParams['fov'],
    aspect: projParams['aspect'],
    near: projParams['near'],
    far: projParams['far'],
  );
}

/// Project flat coordinates [xyz] to pixels on screen given the [pixelProjMatrix].
Vector4 worldToPixels(Vector3 xyz, Matrix4 pixelProjMatrix) {
  final x = xyz[0], y = xyz[1], z = xyz[2];
  assert(x.isFinite && y.isFinite && z.isFinite);

  return transformVector(Vector4(x, y, z, 1), pixelProjMatrix);
}

/// Unproject [xyz] pixels on screen to flat coordinates given the [pixelUnprojMatrix].
Vector pixelsToWorld(Vector3 xyz, Matrix4 pixelUnprojMatrix, {double targetZ}) {
  final x = xyz[0], y = xyz[1], z = xyz[2];
  assert(x.isFinite && y.isFinite, 'invalid pixel coordinate');

  if (z.isFinite) {
    final coord = transformVector(Vector4(x, y, z, 1), pixelUnprojMatrix);
    return coord;
  }

  /// unproject two points to get a line and then find the point on that line with z=0
  final coord0 = transformVector(Vector4(x, y, 0, 1), pixelUnprojMatrix);
  final coord1 = transformVector(Vector4(x, y, 1, 1), pixelUnprojMatrix);

  final z0 = coord0[2], z1 = coord1[2];
  final t = z0 == z1 ? 0 : ((targetZ ?? 0) - z0) / (z1 - z0);

  /// lerp
  final ax = coord0[0], ay = coord0[1];
  return Vector2(ax + t * (coord1[0] - ax), ay + t * (coord1[1] - ay));
}


lib/geo_utils.dart
part of web_mercator;

/// Average circumference (40075 km equatorial, 40007 km meridional)
const EARTH_CIRCUMFERENCE = 40.03e6;

/// Mapbox default altitude
const DEFAULT_ALTITUDE = 1.5;

/// Default size of a map tile
const TILE_SIZE = 512;

/// Convert a distance measurement (assuming a spherical Earth) from a real-world unit into radians.
/// see: [https://github.com/Turfjs/turf/blob/master/packages/turf-helpers/index.ts](turf/helpers)
num lengthToRadians(num distance) {
  const earthRadius = 6371008.8;
  final factor = earthRadius / 1000;
  return distance / factor;
}

/// Takes a [lng, lat] location and calculates the location of a destination point given a [distance] in kilometers; and [bearing] in degrees.
/// see: [https://github.com/Turfjs/turf/blob/master/packages/turf-destination/index.ts](turf/destination)
Vector2 destination(double lng, double lat, {@required num distance, num bearing = 0}) {
  assert(distance.isFinite && distance > 0, 'distance is invalid');

  final lng1 = lng * degrees2Radians;
  final lat1 = lat * degrees2Radians;
  final bearingRad = bearing * degrees2Radians;
  final radians = lengthToRadians(distance);

  final lat2 = asin(sin(lat1) * cos(radians) + cos(lat1) * sin(radians) * cos(bearingRad));
  final lng2 = lng1 + atan2(sin(bearingRad) * sin(radians) * cos(lat1), cos(radians) - sin(lat1) * sin(lat2));

  return Vector2(lng2 * radians2Degrees, lat2 * radians2Degrees);
}

/// Takes a list of [coordinates] and returns a bounding box enclosing them (in [minX, minY, maxX, maxY] order).
/// see: [https://github.com/Turfjs/turf/blob/master/packages/turf-bbox/index.ts](turf/bbox)
List<double> bbox(List<List<double>> coordinates) {
  assert(coordinates != null);
  var result = <double>[double.infinity, double.infinity, double.negativeInfinity, double.negativeInfinity];

  for (final coord in coordinates) {
    if (result[0] > coord[0]) {
      result[0] = coord[0];
    }
    if (result[1] > coord[1]) {
      result[1] = coord[1];
    }
    if (result[2] < coord[0]) {
      result[2] = coord[0];
    }
    if (result[3] < coord[1]) {
      result[3] = coord[1];
    }
  }

  return result;
}

/// Project [lng, lat] on sphere on 512*512 Mercator Zoom 0 tile.
Vector2 lngLatToWorld(double lng, double lat) {
  assert(lng.isFinite);
  assert(lat.isFinite && lat >= -90 && lat <= 90, 'invalid latitude');

  final lambda2 = lng * degrees2Radians;
  final phi2 = lat * degrees2Radians;
  final x = (TILE_SIZE * (lambda2 + pi)) / (2 * pi);
  final y = (TILE_SIZE * (pi + log(tan(PI_4 + phi2 * .5)))) / (2 * pi);

  return Vector2(x, y);
}

/// Unproject world point [x, y] on map onto [lng, lat] on sphere.
Vector2 worldToLngLat(double x, double y) {
  final lambda2 = (x / TILE_SIZE) * (2 * pi) - pi;
  final phi2 = 2 * (atan(exp((y / TILE_SIZE) * (2 * pi) - pi)) - PI_4);

  return Vector2(lambda2 * radians2Degrees, phi2 * radians2Degrees);
}

/// Return the zoom level that gives a 1 meter pixel at a certain [lat].
double getMeterZoom(double lat) {
  assert(lat.isFinite);

  final latCosine = cos(lat * degrees2Radians);

  return scaleToZoom(EARTH_CIRCUMFERENCE * latCosine) - 9;
}

/// Calculate distance scales in meters around current [lat, lng], for both degrees and pixels.
Map<String, List> getDistanceScales(double lng, double lat, {bool highPrecision = false}) {
  assert(lng.isFinite && lat.isFinite);

  final latCosine = cos(lat * degrees2Radians);

  /// Number of pixels occupied by one degree lng around current lat/lon:
  const unitsPerDegreeX = TILE_SIZE / 360;
  final unitsPerDegreeY = unitsPerDegreeX / latCosine;

  /// Number of pixels occupied by one meter around current lat/lon:
  final altUnitsPerMeter = TILE_SIZE / EARTH_CIRCUMFERENCE / latCosine;

  final result = <String, List>{};
  result['unitsPerMeter'] = <double>[altUnitsPerMeter, altUnitsPerMeter, altUnitsPerMeter];
  result['metersPerUnit'] = <double>[1 / altUnitsPerMeter, 1 / altUnitsPerMeter, 1 / altUnitsPerMeter];
  result['unitsPerDegree'] = <double>[unitsPerDegreeX, unitsPerDegreeY, altUnitsPerMeter];
  result['degreesPerUnit'] = <double>[1 / unitsPerDegreeX, 1 / unitsPerDegreeY, 1 / altUnitsPerMeter];

  if (highPrecision) {
    final latCosine2 = (degrees2Radians * tan(lat * degrees2Radians)) / latCosine;
    final unitsPerDegreeY2 = (unitsPerDegreeX * latCosine2) * .5;
    final altUnitsPerDegree2 = (TILE_SIZE / EARTH_CIRCUMFERENCE) * latCosine2;
    final altUnitsPerMeter2 = (altUnitsPerDegree2 / unitsPerDegreeY) * altUnitsPerMeter;

    result['unitsPerDegree2'] = <double>[.0, unitsPerDegreeY2, altUnitsPerDegree2];
    result['unitsPerMeter2'] = <double>[altUnitsPerMeter2, .0, altUnitsPerMeter2];
  }

  return result;
}

/// Calculates camera projection for
Map<String, num> getProjParameters(
  num width,
  num height, {
  double altitude = DEFAULT_ALTITUDE,
  double pitch = 0,
  double nearZMultiplier = 1,
  double farZMultiplier = 1,
}) {
  /// Find the distance from the center point to the center top in altitude units using law of sines.
  final pitchRadians = pitch * degrees2Radians;
  final halfFov = atan(.5 / altitude);
  final topHalfSurfaceDistance = (sin(halfFov) * altitude) / sin(PI_2 - pitchRadians - halfFov);

  /// Calculate z value of the farthest fragment that should be rendered.
  final farZ = cos(PI_2 - pitchRadians) * topHalfSurfaceDistance + altitude;

  return {
    'fov': 2 * halfFov,
    'aspect': width / height,
    'focalDistance': altitude,
    'near': nearZMultiplier,
    'far': farZ * farZMultiplier,
  };
}

/// Returns map settings {lng, lat, zoom} containing the provided [bounds] within the provided [width] & [height].
/// Only supports non-perspective mode.
Map<String, num> fitBounds({
  @required num width,
  @required num height,
  @required List<num> bounds,
  double minExtent = 0,
  double maxZoom = 24,
  dynamic padding = 0,
  List<num> offset = const [0, 0],
}) {
  assert(bounds.length == 4);
  assert(offset.length == 2);

  final west = bounds[0], south = bounds[1], east = bounds[2], north = bounds[3];

  if (padding is int) {
    final num p = padding;
    padding = {'top': p, 'right': p, 'bottom': p, 'left': p};
  } else {
    assert(padding is Map<String, num>);
    assert(padding['top'] is num && padding['right'] is num && padding['bottom'] is num && padding['left'] is num);
  }

  final viewport = MercatorViewport(width: width, height: height);

  final nw = viewport.project(Vector2(west, north)) as Vector2;
  final se = viewport.project(Vector2(east, south)) as Vector2;

  /// width/height on the Web Mercator plane
  final size = <num>[
    max((se[0] - nw[0]).abs(), minExtent),
    max((se[1] - nw[1]).abs(), minExtent),
  ];

  final targetSize = <num>[
    width - padding['left'] - padding['right'] - offset[0].abs() * 2,
    height - padding['top'] - padding['bottom'] - offset[1].abs() * 2,
  ];

  assert(targetSize[0] > 0 && targetSize[1] > 0);

  /// scale = screen pixels per unit on the Web Mercator plane
  final scaleX = targetSize[0] / size[0];
  final scaleY = targetSize[1] / size[1];

  /// Find how much we need to shift the center
  final double offsetX = (padding['right'] - padding['left']) * .5 / scaleX;
  final double offsetY = (padding['bottom'] - padding['top']) * .5 / scaleY;

  final center = Vector3((se[0] + nw[0]) * .5 + offsetX, (se[1] + nw[1]) * .5 + offsetY, double.nan);
  final centerLngLat = viewport.unproject(center) as Vector3;
  final zoom = min(maxZoom, viewport.zoom + log2(min(scaleX, scaleY)).abs());

  assert(zoom.isFinite);
  return {'lng': centerLngLat[0], 'lat': centerLngLat[1], 'zoom': zoom};
}

/// Offset a [lngLatZ] position by meterOffset (northing, easting) [xyz].
Vector3 addMetersToLngLat(Vector3 lngLatZ, Vector3 xyz) {
  final lng = lngLatZ[0], lat = lngLatZ[1], z0 = lngLatZ[2];
  final x = xyz[0], y = xyz[1], z = xyz[2];

  final distanceScales = getDistanceScales(lng, lat, highPrecision: true);
  final unitsPerMeter = distanceScales['unitsPerMeter'];
  final unitsPerMeter2 = distanceScales['unitsPerMeter2'];

  final worldspace = lngLatToWorld(lng, lat);
  worldspace[0] += x * (unitsPerMeter[0] + unitsPerMeter2[0] * y);
  worldspace[1] += y * (unitsPerMeter[1] + unitsPerMeter2[1] * y);

  final newLngLat = worldToLngLat(worldspace[0], worldspace[1]);
  final newZ = (z0 ?? 0) + (z ?? 0);

  return Vector3(newLngLat[0], newLngLat[1], newZ);
}
